### DISCUSSION TITLE
Is this a good way of including various file formats as `Content` in a WebForms project?

### DISCUSSION THREAD

**Discussion by julealgon** [CZEMacLeod/MSBuild.SDK.SystemWeb#84] [Q&A]
I'm migrating some of our legacy deployment processes for a WebForms app. Our current deployment procedure is to just copy the entire projects folder after build, which ends up including a ton of unnecessary files (such as the `.cs` files themselves) in the result, and doesn't support the XML transformation step that happens only on publish. 

I'm now switching into a more standard approach, where we first publish our project using the `/p:DeployOnBuild=true` MSBuild setting, and then grab the output from that.

Because of this change, I need to make sure that all the "content" files (images, javascript, css, etc) in the project are properly considered as `Content` in MSBuild and subsequently copied to the publish folder.

To achieve this, I took a look at how the standard files are added in the SDK, and tried to do something similar for my own needs.

This is what I added to my `csproj` file:
```xml
  <PropertyGroup>
    <DefaultItemExcludes>$(DefaultItemExcludes);**\*.css;*.css</DefaultItemExcludes>
    <DefaultItemExcludes>$(DefaultItemExcludes);**\*.csv;*.csv</DefaultItemExcludes>
    <DefaultItemExcludes>$(DefaultItemExcludes);**\*.gif;*.gif</DefaultItemExcludes>
    <DefaultItemExcludes>$(DefaultItemExcludes);**\*.htm;*.htm</DefaultItemExcludes>
    <DefaultItemExcludes>$(DefaultItemExcludes);**\*.html;*.html</DefaultItemExcludes>
    <DefaultItemExcludes>$(DefaultItemExcludes);**\*.ico;*.ico</DefaultItemExcludes>
    <DefaultItemExcludes>$(DefaultItemExcludes);**\*.jpeg;*.jpeg</DefaultItemExcludes>
    <DefaultItemExcludes>$(DefaultItemExcludes);**\*.jpg;*.jpg</DefaultItemExcludes>
    <DefaultItemExcludes>$(DefaultItemExcludes);**\*.js;*.js</DefaultItemExcludes>
    <DefaultItemExcludes>$(DefaultItemExcludes);**\*.less;*.less</DefaultItemExcludes>
    <DefaultItemExcludes>$(DefaultItemExcludes);**\*.pdn;*.pdn</DefaultItemExcludes>
    <DefaultItemExcludes>$(DefaultItemExcludes);**\*.png;*.png</DefaultItemExcludes>
    <DefaultItemExcludes>$(DefaultItemExcludes);**\*.svg;*.svg</DefaultItemExcludes>
  </PropertyGroup>

  <ItemGroup>
    <Content Include="**\*.css" Exclude="$(DefaultWebFormsItemExcludes)" />
    <Content Include="**\*.csv" Exclude="$(DefaultWebFormsItemExcludes)" />
    <Content Include="**\*.gif" Exclude="$(DefaultWebFormsItemExcludes)" />
    <Content Include="**\*.htm" Exclude="$(DefaultWebFormsItemExcludes)" />
    <Content Include="**\*.html" Exclude="$(DefaultWebFormsItemExcludes)" />
    <Content Include="**\*.ico" Exclude="$(DefaultWebFormsItemExcludes)" />
    <Content Include="**\*.jpeg" Exclude="$(DefaultWebFormsItemExcludes)" />
    <Content Include="**\*.jpg" Exclude="$(DefaultWebFormsItemExcludes)" />
    <Content Include="**\*.js" Exclude="$(DefaultWebFormsItemExcludes)" />
    <Content Include="**\*.less" Exclude="$(DefaultWebFormsItemExcludes)" />
    <Content Include="**\*.pdn" Exclude="$(DefaultWebFormsItemExcludes)" />
    <Content Include="**\*.png" Exclude="$(DefaultWebFormsItemExcludes)" />
    <Content Include="**\*.svg" Exclude="$(DefaultWebFormsItemExcludes)" />
  </ItemGroup>
```

I'd like to check if this is a good approach to the problem. It appears to work just fine, but I'm no MSBuild expert and was wondering if there is a simpler way to achieve the same results.

It looks verbose to me and has some duplication, but it seems better than individually marking each and every content file as `Content` in Visual Studio.

Any feedback is welcome.

**Comment by CZEMacLeod** [comment]
I don't think you need the `DefaultItemExcludes` `PropertyGroup` at all. 
I think the `Content` `Exclude` should be `$(DefaultItemExcludes)` not `$(DefaultWebFormsItemExcludes)` - this prevents it picking up content from folders like node_modules etc. - see #19 
Generally look at the SDK project documentation on globbing etc. as it should all work the same as a modern csproj (or vbproj).
Each project is different, so I can't really advise other than if you have specific issues.
And do you really want to include *.pdn files? Your site publishes Paint.Net files as content?
Remember you will need to either resigister them as static files and add a mime-type, or if they are converted using a handler, register an appropriate handler. If you have contect in e.g. App_Data you can just
```xml
<ItemGroup>
    <Content Include="App_Data\**\*.*" />
</ItemGroup>
```

**Reply by julealgon** [reply]
> I don't think you need the `DefaultItemExcludes` `PropertyGroup` at all.

Aren't the files added to the `None` build action otherwise? If I manually change a file in the project to `Content`, Visual Studio adds a `<None Remove` and a `<Content Include` tag for that item. I thought the `DefaultItemExcludes` was doing that part of the work (removing from `None`). 

If you could elaborate a bit on how this works that would be appreciated. Another reason I did it this way is because this is how it is setup for the webforms files in the SDK:

https://github.com/CZEMacLeod/MSBuild.SDK.SystemWeb/blob/add70be345fd8a988c3a659bd9b482a6988087fd/src/MSBuild.SDK.SystemWeb/Sdk/MSBuild.SDK.SystemWeb.DefaultItems.targets#L3-L10



> I think the `Content` `Exclude` should be `$(DefaultItemExcludes)` not `$(DefaultWebFormsItemExcludes)` - this prevents it picking up content from folders like node_modules etc. - see #19 Generally look at the SDK project documentation on globbing etc. as it should all work the same as a modern csproj (or vbproj).

That was again based on the SDK code. I thought I was doing conceptually the same thing, so I just copied the same approach from there:

https://github.com/CZEMacLeod/MSBuild.SDK.SystemWeb/blob/add70be345fd8a988c3a659bd9b482a6988087fd/src/MSBuild.SDK.SystemWeb/Sdk/MSBuild.SDK.SystemWeb.DefaultItems.props#L69-L78

Does it not apply then?

> And do you really want to include *.pdn files? Your site publishes Paint.Net files as content? Remember you will need to either resigister them as static files and add a mime-type, or if they are converted using a handler, register an appropriate handler. 

I was surprised when I saw that file extension on WinDiff as well when double checking my publish pipeline 😅. We have one of those in one of our "Images" folder. I've since removed it from that list and added the whole "Images" folder as content (along with a few others).

> If you have contect in e.g. App_Data you can just
> 
> ```
> <ItemGroup>
>     <Content Include="App_Data\**\*.*" />
> </ItemGroup>
> ```

Yeah I realized this after posting the question. I changed the approach slightly to explicitly mark some folders as content, and kept only generic extensions that would appear almost everywhere in the list. Something like:

```xml
  <PropertyGroup>
    <DefaultItemExcludes>$(DefaultItemExcludes);**\*.css;*.css</DefaultItemExcludes>
    <DefaultItemExcludes>$(DefaultItemExcludes);**\*.css.map;*.css.map</DefaultItemExcludes>
    <DefaultItemExcludes>$(DefaultItemExcludes);**\*.csv;*.csv</DefaultItemExcludes>
    <DefaultItemExcludes>$(DefaultItemExcludes);**\*.gif;*.gif</DefaultItemExcludes>
    <DefaultItemExcludes>$(DefaultItemExcludes);**\*.htm;*.htm</DefaultItemExcludes>
    <DefaultItemExcludes>$(DefaultItemExcludes);**\*.html;*.html</DefaultItemExcludes>
    <DefaultItemExcludes>$(DefaultItemExcludes);**\*.jpg;*.jpg</DefaultItemExcludes>
    <DefaultItemExcludes>$(DefaultItemExcludes);**\*.js;*.js</DefaultItemExcludes>
    <DefaultItemExcludes>$(DefaultItemExcludes);**\*.less;*.less</DefaultItemExcludes>
    <DefaultItemExcludes>$(DefaultItemExcludes);**\*.png;*.png</DefaultItemExcludes>
    <DefaultItemExcludes>$(DefaultItemExcludes);**\*.xml;*.xml</DefaultItemExcludes>
  </PropertyGroup>

  <ItemGroup>
    <Content Include="**\*.css" Exclude="$(DefaultWebFormsItemExcludes)" />
    <Content Include="**\*.css.map" Exclude="$(DefaultWebFormsItemExcludes)" />
    <Content Include="**\*.csv" Exclude="$(DefaultWebFormsItemExcludes)" />
    <Content Include="**\*.gif" Exclude="$(DefaultWebFormsItemExcludes)" />
    <Content Include="**\*.htm" Exclude="$(DefaultWebFormsItemExcludes)" />
    <Content Include="**\*.html" Exclude="$(DefaultWebFormsItemExcludes)" />
    <Content Include="**\*.jpg" Exclude="$(DefaultWebFormsItemExcludes)" />
    <Content Include="**\*.js" Exclude="$(DefaultWebFormsItemExcludes)" />
    <Content Include="**\*.less" Exclude="$(DefaultWebFormsItemExcludes)" />
    <Content Include="**\*.png" Exclude="$(DefaultWebFormsItemExcludes)" />
    <Content Include="**\*.xml" Exclude="$(DefaultWebFormsItemExcludes)" />
  </ItemGroup>

  <ItemGroup>
    <Content Include="Images\**\*" Exclude="$(DefaultWebFormsItemExcludes)" />
    <Content Include="Images_new\**\*" Exclude="$(DefaultWebFormsItemExcludes)" />
    <Content Include="Includes\**\*" Exclude="$(DefaultWebFormsItemExcludes)" />
    <Content Include="Scripts\**\*" Exclude="$(DefaultWebFormsItemExcludes)" />
    <Content Include="sharedassets\**\*" Exclude="$(DefaultWebFormsItemExcludes)" />
    <Content Include="Public\ExcelTemplates\**\*" Exclude="$(DefaultWebFormsItemExcludes)" />
    <Content Include="Public\PdfTemplates\**\*" Exclude="$(DefaultWebFormsItemExcludes)" />
  </ItemGroup>
```

Unfortunately, we don't have all of our assets centralized in a single folder, which would make this easier to manage here.

**Reply by CZEMacLeod** [reply]
There are quite a few nuances here - but it looks like what you have is working for you so I would go with it.
The `DefaultWebFormsItemExcludes` prevents it picking up items copied to obj/ and bin/ - e.g. published files as well as files from package managers.
It shouldn't normally matter that an item is in None and Content, hence the bit about DefaultItemExcludes.

**Reply by leusbj** [reply]
As @CZEMacLeod mentions there are a few nuances to how the ItemGroups work, and I am hesitant to steer you away from something that is doing what you need to do, but I will share this as a potential starting point for your adventures/experiments in ItemGroups
```
  <ItemGroup>
    <None Remove="**/*" 
          Label="Remove all files from the None ItemGroup, since we are going to have most things be content" />
    <Content Include="**/*" 
             Exclude="$(DefaultItemExcludes);
             $(DefaultExcludesInProjectFolder);
             $(AppDesignerFolder)/**/*;
             **/*$(DefaultLanguageSourceExtension);
             **/*.resx"  
            Label="Add Most Items under the project to the Content ItemGroup except compile files, resx files, and files under the bin, obj, and 'Properties/MyProject' directories"/>
    <None Include="$(AppDesignerFolder)/**/*" 
             Exclude="$(DefaultItemExcludes);
             $(DefaultExcludesInProjectFolder);
             **/*$(DefaultLanguageSourceExtension);
             **/*.resx"  
          Label="Add back to the None ItemGroup the items from the 'Properties/MyProject' directory " />
  </ItemGroup>

```
